File: packages/sdk/python/caracalai_sdk/advanced.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

Advanced surface: low-level primitives, codec, ambient context plumbing, and
the raw coordinator client. Most integrators only need ``caracalai_sdk``;
reach for these when building a transport adapter or framework shim.
"""

from .envelope import (
    HEADER_AUTHORIZATION,
    HEADER_BAGGAGE,
    HEADER_TRACEPARENT,
    BAGGAGE_AGENT_SESSION,
    BAGGAGE_DELEGATION_EDGE,
    BAGGAGE_HOP,
    BAGGAGE_PARENT_EDGE,
    MAX_HOP,
    Envelope,
    decode_envelope,
    encode_envelope,
    encode_baggage,
    extract,
    format_traceparent,
    from_headers,
    inject,
    parse_baggage,
    parse_traceparent,
    to_headers,
)
from .context import (
    CaracalContext,
    abind,
    bind,
    current,
    from_envelope,
    to_envelope,
    try_current,
    with_overrides,
)
from .coordinator import (
    AgentKind,
    CoordinatorClient,
    DelegationRequest,
    DelegationResponse,
    SpawnRequest,
    SpawnResponse,
    create_delegation,
    spawn_agent,
    terminate_agent,
)
from .primitives import with_agent, with_delegation
from .client import Caracal, CaracalConfig
from .http import CaracalASGIMiddleware

__all__ = [
    "HEADER_AUTHORIZATION",
    "HEADER_TRACEPARENT",
    "HEADER_BAGGAGE",
    "BAGGAGE_AGENT_SESSION",
    "BAGGAGE_DELEGATION_EDGE",
    "BAGGAGE_PARENT_EDGE",
    "BAGGAGE_HOP",
    "MAX_HOP",
    "Envelope",
    "decode_envelope",
    "encode_envelope",
    "encode_baggage",
    "format_traceparent",
    "from_headers",
    "to_headers",
    "inject",
    "extract",
    "parse_baggage",
    "parse_traceparent",
    "CaracalContext",
    "current",
    "try_current",
    "bind",
    "abind",
    "with_overrides",
    "to_envelope",
    "from_envelope",
    "AgentKind",
    "CoordinatorClient",
    "SpawnRequest",
    "SpawnResponse",
    "DelegationRequest",
    "DelegationResponse",
    "spawn_agent",
    "terminate_agent",
    "create_delegation",
    "with_agent",
    "with_delegation",
    "Caracal",
    "CaracalConfig",
    "CaracalASGIMiddleware",
]

File: packages/sdk/ts/src/advanced.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * Advanced surface: low-level primitives, codec, bound context plumbing,
 * and the raw coordinator client. Most integrators only need the default
 * "@caracalai/sdk" entrypoint; reach for these when building a transport
 * adapter or framework integration.
 */

export {
  HeaderAuthorization,
  HeaderTraceparent,
  HeaderTracestate,
  HeaderBaggage,
  BaggageAgentSession,
  BaggageDelegationEdge,
  BaggageParentEdge,
  BaggageSession,
  BaggageHop,
  MaxHop,
  formatTraceparent,
  parseTraceparent,
  encodeBaggage,
  parseBaggage,
  fromHeaders,
  decodeEnvelope,
  encodeEnvelope,
  toHeaders,
} from './envelope.js'
export type { Envelope, HeaderGetter, HeaderSetter } from './envelope.js'
export {
  current,
  captureContext,
  bind,
  withOverrides,
  toEnvelope,
  fromEnvelope,
  fromVerifiedEnvelope,
  describeAuthority,
} from './context.js'
export type { CaracalContext, AuthoritySummary, VerifiedClaims } from './context.js'
export {
  CoordinatorError,
  Lifecycle,
  startCoordinatorSession,
  terminateSession,
  acquireSessionLease,
  heartbeatSession,
  createDelegation,
  revokeDelegation,
  getInboundDelegation,
  listInboundDelegations,
} from './coordinator.js'
export type {
  CoordinatorCallEvent,
  CoordinatorClient,
  SessionStatus,
  DelegationConstraints,
  StartSessionRequest,
  StartSessionResponse,
  DelegationRequest,
  DelegationResponse,
  HeartbeatResponse,
  InboundDelegation,
} from './coordinator.js'
export { Authority, session, delegate, acceptDelegation, startSession, attachSession } from './primitives.js'
export type {
  AuthorityMode,
  SessionInput,
  DelegateInput,
  Delegation,
  StartSessionInput,
  SessionHandle,
  AttachSessionInput,
} from './primitives.js'
export { caracalContextMiddleware, caracalFastifyHook } from './http.js'
export type { IncomingLike, FastifyRequestLike, ServerResponseLike, FastifyReplyLike, ConnectMiddleware } from './http.js'
export {
  Caracal,
  createAdvancedClient,
  createAdvancedClientFromConfig,
  createAdvancedClientFromCredentials,
  createAdvancedClientFromEnv,
} from './client.js'
export type {
  CaracalConfig,
  SessionOptions,
  DelegateOptions,
  LifecycleHook,
  CallOptions,
  TokenSource,
  ClientCredentials,
  CredentialsResolver,
  ClientSecretOptions,
} from './client.js'
